lnwallet/wallet.go

```diff
@@ -1485,3 +1485,93 @@ func coinSelect(feeRate SatPerKWeight, amt btcutil.Amount,
 	}
 }
+
+// coinSelectSubtractFees attempts to select coins such that we'll spend up to
+// amt in total after fees, adhering to the specified fee rate. The selected
+// coins, the final output and change values are returned.
+func coinSelectSubtractFees(feeRate SatPerKWeight, amt,
+	dustLimit btcutil.Amount, coins []*Utxo) ([]*Utxo, btcutil.Amount,
+	btcutil.Amount, error) {
+
+	// First perform an initial round of coin selection to estimate
+	// the required fee.
+	totalSat, selectedUtxos, err := selectInputs(amt, coins)
+	if err != nil {
+		return nil, 0, 0, err
+	}
+
+	var weightEstimate input.TxWeightEstimator
+	for _, utxo := range selectedUtxos {
+		switch utxo.AddressType {
+		case WitnessPubKey:
+			weightEstimate.AddP2WKHInput()
+		case NestedWitnessPubKey:
+			weightEstimate.AddNestedP2WKHInput()
+		default:
+			return nil, 0, 0, fmt.Errorf("Unsupported "+
+				"address type: %v", utxo.AddressType)
+		}
+	}
+
+	// Channel funding multisig output is P2WSH.
+	weightEstimate.AddP2WSHOutput()
+
+	// At this point we've got two possibilities, either create a
+	// change output, or not. We'll first try without creating a
+	// change output.
+	//
+	// Estimate the fee required for a transaction without a change
+	// output.
+	totalWeight := int64(weightEstimate.Weight())
+	requiredFee := feeRate.FeeForWeight(totalWeight)
+
+	// For a transaction without a change output, we'll let everything go
+	// to our multi-sig output after subtracting fees.
+	outputAmt := totalSat - requiredFee
+	changeAmt := btcutil.Amount(0)
+
+	// If the the output is too small after subtracting the fee, the coin
+	// selection cannot be performed with an amount this small.
+	if outputAmt <= dustLimit {
+		return nil, 0, 0, fmt.Errorf("output amount(%v) after "+
+			"subtracting fees(%v) below dust limit(%v)", outputAmt,
+			requiredFee, dustLimit)
+	}
+
+	// We were able to create a transaction with no change from the
+	// selected inputs. We'll remember the resulting values for
+	// now, while we try to add a change output. Assume that change output
+	// is a P2WKH output.
+	weightEstimate.AddP2WKHOutput()
+
+	// Now that we have added the change output, redo the fee
+	// estimate.
+	totalWeight = int64(weightEstimate.Weight())
+	requiredFee = feeRate.FeeForWeight(totalWeight)
+
+	// For a transaction with a change output, everything we don't spend
+	// wil go to change.
+	newChange := totalSat - amt
+	newOutput := amt - requiredFee
+
+	// If adding a change output lead to both outputs being above
+	// the dust limit, we'll add the change ouput. Otherwise we'll
+	// go with the no change tx we originally found.
+	if newChange > dustLimit && newOutput > dustLimit {
+		outputAmt = newOutput
+		changeAmt = newChange
+	}
+
+	// Sanity check the resulting ouput values to make sure we
+	// don't burn a great part to fees.
+	totalOut := outputAmt + changeAmt
+	fee := totalSat - totalOut
+
+	// Fail if more than 20% goes to fees.
+	if fee > totalOut/5 {
```

Comment:
Possibly make this 20% value a config parameter? In certain situations like testnet or simnet, the user may not care.

Reply:
Are you thinking a user configurable setting? 

Since the user can already control fees for manually opened channels, maybe it makes sense to make it an `autopilot` setting?

Reply:
agree on making this check configurable! however i don't think a percentage makes much sense here, since the fees paid are independent of the amount being sent. 20% might make sense for $1 channel, but not a $100 or $1000. perhaps this is better expressed as a max fee rate?

Reply:
I added the check in the first place to avoid cases where a user would deposit say $5 into the wallet, and autopilot opened a channel where $3 went to fees. Setting a max fee rate in that scenario wouldn't help, since $3 fee might be okay if you are opening large channels.

We are targeting app users primarily here, so maybe we should discuss it in that context.

Reply:
> Setting a max fee rate in that scenario wouldn't help, since $3 fee might be okay if you are opening large channels.

Why doesn't it help? To me it seems more effective than setting a percentage, since we're directly limiting how much we allocate to fees rather than assuming people are okay paying higher fees simply because they have more money.

> We are targeting app users primarily here, so maybe we should discuss it in that context.

We should still be cognizant of non-app users of autopilot or developers using lnd+autopilot in their own applications, of which there are many on mainnet.

Reply:
> Why doesn't it help?

Because it's absolute. In that "deposit $5, but get $3" scenario, what max fee should be set there? As this example is denominated in fiat, one would also need to factor in the current exchange rate as well. Alternatively, we could have "fee bands", kinda like a piece-wise function to decide what the proper % limit should be given the channel size. So for a 30k satoshi channel, 10k satoshis may be unacceptable as a total fee, while for a 300k satoshi channel 10k satoshis isn't so bad.

Reply:
> Because it's absolute.

Not sure I follow, aren't both absolute? One is don't go above this %, the other is don't go above this sat/byte.

In that "deposit $5, but get $3" scenario, what max fee should be set there? As this example is denominated in fiat, one would also need to factor in the current exchange rate as well.

Was thinking that a user would just tell autopilot to never use more than, e.g., 5 sat/byte for funding. Exchange rates wouldn't come into play at all

> Alternatively, we could have "fee bands", kinda like a piece-wise function to decide what the proper % limit should be given the channel size. So for a 30k satoshi channel, 10k satoshis may be unacceptable as a total fee, while for a 300k satoshi channel 10k satoshis isn't so bad.

To me it'd be less intuitive to specify a number of satoshis, because then you have to convert into dollar amounts. When people see a fee rate, they now roughly know the difference between a good and a bad sat/byte value. Perhaps fee rate bands? :)

Reply:
I don't think this question has a "right" answer, and should be getting more discussion, at least before we start implementing a more complex heuristic.

I'm okay just removing this sanity check for now, even though app users might end up paying high percentage wise fees for opening channels (earlier we had this capped as a result of targeting 95% autopilot allocation in the app, which would leave only 5% to fees). Or just set it at 50% or something, with a TODO of improving it.

Reply:
@halseth sounds good to me, in that case i'd just leave the 20% heuristic with a todo

Reply:
TODO added